### PR TITLE
Add line to Tuple of Ast

### LINKED ISSUE #145: Add line number to Tuple

### PR DESCRIPTION
**Pull request by kmizu**
Fix #145

### CODE REVIEW

**Comment by kmizu** [comment]
Could you mind review this pull request ? @yoshihiro503 and/or @amutake

File: lib/ast.ml [hunk 1]
```diff
@@ -33,6 +33,21 @@ and pattern' =
     | PatNil
 [@@deriving sexp_of]
 
+let line_number_of_t = (function
```
Comment (amutake) [inline lib/ast.ml]: nits: parens are not needed.

```ocaml
let line_number_of_t = function
| Constant (line, _) -> line
...
| ListNil -> -1

let string_of_t t = ...
```

is valid syntax.

File: lib/derivation.ml [hunk 2]
```diff
@@ -18,7 +18,7 @@ let rec derive context = function
         let line_in_file = line in
         Error Known_error.(FialyzerError (UnboundVariable {filename; line=line_in_file; variable=Var v}))
      end
-  | Tuple exprs ->
+  | Tuple (_line, exprs) ->
```
Comment (amutake) [inline lib/derivation.ml]: Could you use this line number as line of error message in `None` pattern?

File: lib/from_erlang.ml [hunk 3]
```diff
@@ -281,9 +280,19 @@ and function_of_clauses clauses =
       let tuple_pattern = PatTuple ps in
       (((tuple_pattern, Constant (line, Atom ("true"))), expr_of_erlang_expr' body), arity)
     ) |> List.unzip in
+     let line_number_of_clause = (function
```
Comment (amutake) [inline lib/from_erlang.ml]: nits: parens are not needed.

File: lib/from_erlang.ml [hunk 4]
```diff
@@ -281,9 +280,19 @@ and function_of_clauses clauses =
...
+     | ((PatCons (_, _), _), term) -> Ast.line_number_of_t term
+     | ((PatVar _, _), term) -> Ast.line_number_of_t term
+     | ((PatNil, _), term) -> Ast.line_number_of_t term
+     ) in
      let make_fresh_variables length = fill (fun () -> Variable.create()) length |> List.rev in
      let make_case cs fresh_variables =
-       let fresh_tuple = Tuple (fresh_variables |> List.map ~f:(fun v -> Var (-1, v))) in
+       let fresh_tuple = Tuple (
+         line_number_of_clause (List.hd_exn cs),
+         fresh_variables |> List.map ~f:(fun v -> Var (-1, v))
```
Comment (amutake) [inline lib/from_erlang.ml]: I think it is convenient for debugging to specify line number as possible.
So, how about this?

```ocaml
  let l = line_number_of_clause (List.hd_exn cs) in
  let fresh_tuple = Tuple (l, fresh_variables |> List.map ~f:(fun v -> Var (l, v))) in
```

Comment (kmizu) [reply lib/from_erlang.ml]: OK.  I'll do it.

File: lib/ast.ml (hunk 1 above)
Comment (kmizu) [reply lib/ast.ml]: I see.

File: lib/derivation.ml (hunk 2 above)
Comment (kmizu) [reply lib/derivation.ml]: OK.

**Comment by kmizu** [comment]
@amutake I've fixed the points you mentioned.  If there is no problems,  could you do LGTM ?

File: lib/derivation.ml (hunk 2 above)
Comment (amutake) [reply lib/derivation.ml]: Sorry, this is my mistake. It's no problem.

**Comment by kmizu** [comment]
Thanks!

**Pull request merged**